### PR TITLE
[COMCTL32] Fix header redraw glitches when resizing

### PR DESCRIPTION
**Pull request by Werru2955**
## Purpose

- Fix COMCTL32 headers from being garbage text when resizing while drawn off screen.

JIRA issue: [CORE-19404](https://jira.reactos.org/browse/CORE-19404)

## Proposed changes

- Invalidate header item rect before scrolling to prevent the divider from being redrawn over and over.
- Invalidate header item text through adding flag SW_INVALIDATE to ScrollWindowEx call to prevent the text from being redrawn over and over.

Before:
before

After:
after

### CODE REVIEW

File: dll/win32/comctl32/header.c
Comment (cbialorucki) [inline dll/win32/comctl32/header.c:1951]: Unfortunately as of now, this code is still synced with Wine so your changes must be guarded using `#ifdef` and `#ifndef` statements.

```suggestion
#ifdef __REACTOS__
                    InvalidateRect(infoPtr->hwndSelf, &lpItem->rect, FALSE);
#endif
```

File: dll/win32/comctl32/header.c
Comment (cbialorucki) [inline dll/win32/comctl32/header.c:1957]: Similar to the suggestion above, guard these changes with `#ifdef __REACTOS__`.
```suggestion
#ifdef __REACTOS__
                     ScrollWindowEx(infoPtr->hwndSelf, nWidth - nOldWidth, 0, &rcScroll, &rcClient,
                                                NULL, NULL, SW_INVALIDATE);
#else
                    ScrollWindowEx(infoPtr->hwndSelf, nWidth - nOldWidth, 0, &rcScroll, &rcClient, NULL, NULL, 0);
                    InvalidateRect(infoPtr->hwndSelf, &lpItem->rect, FALSE);
#endif
```

**Comment by Werru2955** [comment]
Thanks for the help, committed your added #ifdefs to the PR.

**Comment by binarymaster** [comment]
Please set your full name in GitHub profile (keep it in sync with the one in commits), so your commits done in GitHub would be properly authored.

image

**Comment by Werru2955** [comment]
Added my full name to my GitHub profile. If there's anything else I need to do let me know.

**Review by oleg-dubinskiy — APPROVED** [review]
LGTM for the change itself, if it properly fixes the problem from the linked ticket.
But since this code is taken from (and (ideally) regularly synced to) Wine, and we closely cooperate with Wine project, it would be nice if you send your changes to them too.
As we can see, this code is same in the current Wine master https://source.winehq.org/git/wine.git/blob/d07019e4d174fd1b5d8b74ff2a36e900c07d96d7:/dlls/comctl32/header.c#l1945, so your patch can be a still valid contribution for it.
Later, when the whole code will be synced, we can remove the `__REACTOS__` ifdefs.
Nowadays, to contribute there, you can simply create a report and send a Pull Request to the Wine repository on GitLab, instead of sending your patch via email. See https://wiki.winehq.org/Submitting_Patches for details. :slightly_smiling_face:

**Review by JoachimHenze — APPROVED** [review]

**Comment by oleg-dubinskiy** [comment]
See (currently WIP) #6789 PR too. It synchronizes comctl32 to Wine 5.0.

**Comment by binarymaster** [comment]
> If there's anything else I need to do let me know.

You also need to uncheck `Keep my email addresses private` in [GitHub email settings](https://github.com/settings/emails). It's currently uses a noreply email:

image

Thanks.

**Comment by cbialorucki** [comment]
> Later, when the whole code will be synced, we can remove the `__REACTOS__` ifdefs.

It's very unlikely that Wine will approve this PR, they do not have a maintainer for common controls (or any shell component for that matter)

**Comment by Werru2955** [comment]
Okay, my Email address is now public on GitHub. A bit concerning that wine doesn't have a maintainer for their shell components though,

**Review by cbialorucki — APPROVED** [review]

**Review by DarkFire01 — APPROVED** [review]

**Comment by HBelusca** [comment]
@Werru2955 Could you try also, keeping pressed a header "button" and dragging it too?

File: dll/win32/comctl32/header.c
Comment (HBelusca) [reply dll/win32/comctl32/header.c:1952]: Is this call needed, since there is again an invalidate on the same window below?

**Comment by Werru2955** [comment]
> @Werru2955 Could you try also, keeping pressed a header "button" while dragging it too?

Tried doing that just now, and the behavior is the same as in previous builds, although not correct with what it does on my Windows 10 PC, where it changes the order of the toolbar. I guess that's another commit I could do later, lol.

File: dll/win32/comctl32/header.c
Comment (Werru2955) [reply dll/win32/comctl32/header.c:1952]: For some reason not having that call produces redraw issues in the not text portion of the tab.

https://github.com/reactos/reactos/assets/65135562/ac5e8849-acad-45ff-901f-7e9cd8479dec

**Pull request merged**
